### PR TITLE
[FLINK-31843][runtime] Improve performance of SlotSelectionStrategy by bulk select.

### PR DESCRIPTION
**Pull request by huwh**
## What is the purpose of the change
Improving the performance of the SlotSelectionStrategy by bulk selecting tasks in the same region.

## Brief change log
  - *Introduce FreeSlotCandidatesManager to manage available slots for candidates*
  - *Migerate SlotInfoWithUtilization to FreeSlotCandidatesManager for SlotSelectionStrategy*
  - *PhysicalSlotProvider support bulk select available slots*

## Verifying this change
  - *Added unit tests for FreeSlotCandidatesManager*
  - *Manually verified th change by wordcount job with 20000 parallelism*
  - *Compare the benchmark of handleGlobalFailureAndRestartAllTasks*

The benchmark result of handleGlobalFailureAndRestartAllTasks are:
master branch:
```
Benchmark                              (jobConfiguration)  Mode  Cnt      Score     Error  Units
handleGlobalFailureAndRestartAllTasks               BATCH    ss   20   2625.949 ? 204.485  ms/op
handleGlobalFailureAndRestartAllTasks           STREAMING    ss   20   8417.172 ? 543.077  ms/op
handleGlobalFailureAndRestartAllTasks        BATCH_EVENLY    ss   20   4914.222 ? 375.593  ms/op
handleGlobalFailureAndRestartAllTasks    STREAMING_EVENLY    ss   20  20335.352 ? 894.569  ms/op
```
after this pr 
```
Benchmark                              (jobConfiguration)  Mode  Cnt      Score      Error  Units
handleGlobalFailureAndRestartAllTasks               BATCH    ss   20   2096.568 ?   95.519  ms/op
handleGlobalFailureAndRestartAllTasks           STREAMING    ss   20   1727.778 ?  171.373  ms/op
handleGlobalFailureAndRestartAllTasks        BATCH_EVENLY    ss   20   5117.082 ?  341.998  ms/op
handleGlobalFailureAndRestartAllTasks    STREAMING_EVENLY    ss   20  17622.214 ? 1218.075  ms/op
```
The time cost of handleGlobalFailureAndRestartAllTasks for the STREAMING scene is reduced by 80%, and the BATCH and STREAMING_EVENLY scene is reduced by 20%.

However, BATCH_EVENLY has a slight performance regression (less than 5%). I think this is acceptable. Because: 1) Batch Evenly is a strategy that is unlikely to be used in production, batch tasks run for a short time, and resources can be released when they are finished; 2) This only affects the failover process, and part of batch tasks (with block shuffling) will not trigger the global Failover;

The entire benchmark results run in my local ECS are:
- master branch default benchmarks:  [master-default-jmh-result.csv](https://github.com/apache/flink/files/12136552/master-default-jmh-result.csv)
- master branch scheduler benchmarks: [master-scheduler-jmh-result.csv](https://github.com/apache/flink/files/12136553/master-scheduler-jmh-result.csv)
- changed branch default benchmarks: [slotchange-default-jmh-result.csv](https://github.com/apache/flink/files/12136554/slotchange-default-jmh-result.csv)
- changed branch scheduler benchmarks: [slotchange-scheduler-jmh-result.csv](https://github.com/apache/flink/files/12136555/slotchange-scheduler-jmh-result.csv)
- The results of org.apache.flink.benchmark.KeyByBenchmarks.tupleKeyBy for this report are very different. But it's fine after re-run it multiple times.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`:  (no)
  - The serializers:  (no)
  - The runtime per-record code paths (performance sensitive):  (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature?  (no)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SlotSharingExecutionSlotAllocator.java [hunk 1]
```diff
@@ -225,35 +229,66 @@ private void cancelLogicalSlotRequest(ExecutionVertexID executionVertexId, Throw
         return assignments;
     }
 
-    private SharedSlot getOrAllocateSharedSlot(
-            ExecutionSlotSharingGroup executionSlotSharingGroup,
+    private void assignSharedSlots(
```
[the remark this answers is not in the document]
Comment (KarmaGYZ) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SlotSharingExecutionSlotAllocator.java:232]: I lean to return the SSG set without exiting shared slot. It's not a good practice for me to modify the input argument. Same as other methods.

**Review by KarmaGYZ — COMMENTED** [review]
Thanks for the PR, @huwh . I give it a first pass and left some comments. Please take a look.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/FreeSlotCandidatesManager.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+import org.apache.flink.runtime.jobmaster.SlotInfo;
+import org.apache.flink.util.Preconditions;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/** Manage free slots as candidates, used for {@link SlotSelectionStrategy}. */
+public class FreeSlotCandidatesManager {
```
[the remark this answers is not in the document]
Comment (KarmaGYZ) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/FreeSlotCandidatesManager.java:33]: After an offline discussion with @huwh , we found that request the `TaskExecutorUtilization` by allocationID would badly harm the performance according to the benchmark. Thus we'll keep the current implementation.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/AllocatedSlotPool.java [hunk 3]
```diff
@@ -117,6 +117,13 @@ public interface AllocatedSlotPool {
      */
     Collection<FreeSlotInfo> getFreeSlotsInformation();
 
+    /**
+     * Returns information about all currently free slots.
+     *
+     * @return free slot information
+     */
+    FreeSlotCandidatesManager getFreeSlotsCandidates();
```
[the remark this answers is not in the document]
Comment (KarmaGYZ) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/AllocatedSlotPool.java:125]: @huwh Could you help me to understand why we still need it?

[the remark this answers is not in the document]
Comment (huwh) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/AllocatedSlotPool.java:125]: Sorry for the late reply. There are many other scenarios where they need to get the free slots, such as when the AdaptiveScheduler determining parallelism, and checking whether slot could be released for idle timeout, and so on. In these scenarios, the semantics of Slot is not Candidate, so I kept the original interface `getFreeSlotsInformation`

[the remark this answers is not in the document]
Comment (KarmaGYZ) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/AllocatedSlotPool.java:125]: It's a bit confusing to me. Could you elaborate more on the different of these two methods? 

From my understanding, both of the API return the information of current free slots, exactly the same free slot set. `getFreeSlotsCandidates` only wraps the free slot set with a bookkeeping class `FreeSlotCandidatesManager`(reserve slot inside the manager will not take any effect on the slot state of slotPool though). Regarding the naming "candidate" and "candidateManager", maybe we can replace it with something like "getFreeSlotsInformation" and "FreeSlotInfoTracker".

**Comment by huwh** [comment]
@KarmaGYZ Thanks for review, I have updated this pr, PTAL

Because the class name has changed, it's hard to create fixup commits without conflict, so I change the origin commit and force pushed. The main change is here 
 https://github.com/apache/flink/compare/8d753b63d4984cba13641b7a4652df3c1687c51a..67e21c99d13f6e9277bd39d2a91f7ea5abbd5a0d

After this chan, I rebase it into the master branch.

**Comment by huwh** [comment]
@flinkbot run azure

**Comment by KarmaGYZ** [comment]
@flinkbot run azure

**Comment by KarmaGYZ** [comment]
@flinkbot run azure

**Comment by KarmaGYZ** [comment]
@flinkbot run azure

**Comment by huwh** [comment]
The failed 'Kubernetes test' is caused by [FLINK-32632](https://issues.apache.org/jira/browse/FLINK-32632).

**Pull request merged**